### DISCUSSION TITLE
How to add new variables and equations to an existing model

### DISCUSSION THREAD

**Discussion by nzc22** [pybamm-team/PyBaMM#2876] [Q&A]
What is the general process for adding new variables and equations based on existing models (such as DFN), and changing some of the equations of the original model (such as adding new reactions, the total current density will include the current density of the new reaction)?

Are there any relevant examples?

**Comment by RuiheLi** [comment]
Hi~ The first thing to do may be to install PyBaMM as a developer mode, you can find out how to do that [here](https://pybamm.readthedocs.io/en/latest/source/user_guide/installation/install-from-source.html).
Regarding add new variables and equations, there are plenty of examples like [Tutorial 10 - Creating a model](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/Getting%20Started/Tutorial%2010%20-%20Creating%20a%20model.ipynb), [Tutorial 11 - Creating a submodel](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/Getting%20Started/Tutorial%2011%20-%20Creating%20a%20submodel.ipynb), and this [folder](https://github.com/pybamm-team/PyBaMM/tree/develop/examples/notebooks/Creating%20Models).
Also it may be very helpful to get familar with the following source files as they kind of control the overall model hierarchy:
[base_model.py](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/base_model.py), [base_battery_model.py ](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/full_battery_models/base_battery_model.py), [base_lithium_ion_model.py](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/full_battery_models/lithium_ion/base_lithium_ion_model.py)
